Raises error if there is empty cookbook directory at the time o…

Vasu1105 opened the pull request:
Signed-off-by: Vasu1105 <vasundhara.jagdale@msystechnologies.com>

<!--- Provide a short summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail, what problems does it solve? -->
knife cookbook upload was giving nil class error when trying to upload cookbooks and when cookbook repo has empty cookbook directory. This PR fixes the nil class error and raises the error if cookbook repository contains the empty cookbook directory so that user can fix this empty cookbooks and then do the cookbook upload. Also added deprecation warning for load method as we now use load! so that we can raise the error while loading cookbooks. 
## Related Issue
#9010 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

vsingh-msys commented:
I think we should also verify working of `bundle exec knife cookbook upload COOKBOOK_NAME` features working as expected or not.

Also if specific cookbook upload raises only warning we should raise the ERROR in this case?

lamont-granquist commented:
This should be a hard `raise` not a warn.  Cookbooks with no metadata need to just fail and users need to be forced to fix that before continuing.

lamont-granquist commented:
So I'd suggest that the cookbook_version_loader needs to be changed to just throw on an empty cookbook or any cookbook that lacks a metadata.json or metadata.rb or which does not have a name or a version attribute:

https://github.com/chef/chef/blob/bddd4852791df30a78da74fd7e432e7efc87cf54/lib/chef/cookbook/cookbook_version_loader.rb#L177-L183

That code needs to get changed to be a lot more strict and less permissive.

Vasu1105 commented on lib/chef/cookbook_uploader.rb:
```diff
@@ -139,12 +139,14 @@ def uploader_function_for(file, checksum, url, checksums_to_upload)
-        Chef::Log.info("Validating templates")
-        exit(1) unless syntax_checker.validate_templates
-        Chef::Log.info("Syntax OK")
+        unless cb.nil?
```
> So I'd suggest that the cookbook_version_loader needs to be changed to just throw on an empty cookbook or any cookbook that lacks a metadata.json or metadata.rb or which does not have a name or a version attribute:
> 
> https://github.com/chef/chef/blob/bddd4852791df30a78da74fd7e432e7efc87cf54/lib/chef/cookbook/cookbook_version_loader.rb#L177-L183
> 
> That code needs to get changed to be a lot more strict and less permissive.

@lamont-granquist does not upload the blank cookbook.

```$ bundle exec knife cookbook upload blank_dir -c ~/workspace/chef/chef-repo/.chef/knife.rb
WARN: Found a directory blank_dir in the cookbook path, but it contains no cookbook files. skipping.
Uploaded 0 cookbooks.
```

Second in bulk upload if we raise error it will break uploading of other cookbooks so what it is currently doing is skipping the blank cookbook (which does not have any files) to upload by giving warning message and uploading rest.

```
$ bundle exec knife cookbook upload --all -c ~/workspace/chef/chef-repo/.chef/knife.rb
WARN: Found a directory test in the cookbook path, but it contains no cookbook files. skipping.
WARN: Found a directory blank_dir in the cookbook path, but it contains no cookbook files. skipping.
Uploading audit          [5.0.0]
Uploading compat_resource [12.19.1]
Uploading mangled        [0.1.0]
Uploading starter        [0.0.0]
Uploading test_freeze    [0.1.3]
Uploading without_metadata [0.1.0]
Uploaded all cookbooks.
```

This PR is only to fix the nil class error which is coming while uploading bulk cookbooks with one of the cookbook as blank and single blank cookbook without any files.

Vasu1105 commented:
> So I'd suggest that the cookbook_version_loader needs to be changed to just throw on an empty cookbook or any cookbook that lacks a metadata.json or metadata.rb or which does not have a name or a version attribute:
> 
> https://github.com/chef/chef/blob/bddd4852791df30a78da74fd7e432e7efc87cf54/lib/chef/cookbook/cookbook_version_loader.rb#L177-L183
> 
> That code needs to get changed to be a lot more strict and less permissive.


@lamont-granquist does not upload the blank cookbook.

```$ bundle exec knife cookbook upload blank_dir -c ~/workspace/chef/chef-repo/.chef/knife.rb
WARN: Found a directory blank_dir in the cookbook path, but it contains no cookbook files. skipping.
Uploaded 0 cookbooks.
```

Second in bulk upload if we raise error it will break uploading of other cookbooks so what it is currently doing is skipping the blank cookbook (which does not have any files) to upload by giving warning message and uploading rest.

```
$ bundle exec knife cookbook upload --all -c ~/workspace/chef/chef-repo/.chef/knife.rb
WARN: Found a directory test in the cookbook path, but it contains no cookbook files. skipping.
WARN: Found a directory blank_dir in the cookbook path, but it contains no cookbook files. skipping.
Uploading audit          [5.0.0]
Uploading compat_resource [12.19.1]
Uploading mangled        [0.1.0]
Uploading starter        [0.0.0]
Uploading test_freeze    [0.1.3]
Uploading without_metadata [0.1.0]
Uploaded all cookbooks.
```

This PR is only to fix the nil class error which is coming while uploading bulk cookbooks with one of the cookbook as blank and single blank cookbook without any files.

Vasu1105 commented:
@lamont-granquist i have just added deprecation warning for load and copied code of load in load! method and updated code to call load! instead of load. May be we can go through the errors which are coming and decide the approach.

lamont-granquist commented on lib/chef/cookbook/cookbook_version_loader.rb:
```diff
@@ -93,6 +88,26 @@ def load
+      #   if empty?
+      #     Chef::Log.warn "Found a directory #{cookbook_name} in the cookbook path, but it contains no cookbook files. skipping."
+      #   end
+      #   cookbook_settings
```
Ah what you should probably do here is just force it to `raise` instead of warn and hard error out.

I think I can see what its going to do though.  It looks like this is mostly used by the cookbook_version_loader and by `chef_fs/file_system/repository/chef_repository_file_system_cookbook_dir.rb` (I think).  If you switch that to a raise, and then convert those calls from `load` to `load!` I think it'll be close to passing.

lamont-granquist commented on lib/chef/cookbook/cookbook_version_loader.rb:
```diff
@@ -88,12 +75,17 @@ def load
 
-      alias :load_cookbooks :load
+      def load
+        Chef::Log.warn "load method is deprecated. Use load! instead"
```
that should really raise here as well.  also `load_cookbooks` should still be aliased to `load` for this test.

Vasu1105 replied:
@lamont-granquist I have done the changes so that we can see the failing test.

tas50 commented:
@Vasu1105 Can you rebase this and take a look at the test failures

Vasu1105 commented:
> @Vasu1105 Can you rebase this and take a look at the test failures

Sure busy on another PR so once that's done will have a look at these.

Vasu1105 commented:
@lamont-granquist Fixed failing specs and this is ready for review. Please have a look at it once you are back from vacations :)

Vasu1105 commented:
> this one has merge conflicts now probably due to your other one going in.

@lamont-granquist Resolved them could you please have a look.

lamont-granquist commented:
yeah looks like this one needs a re-rebase, i can't convince the kitchen tests to build off of the merge commit with the actual master branch as of today.
